### PR TITLE
use same behaviour for valueOrNull as for valueOrDefault

### LINKED ISSUE #2316: FailFastConfig.valueOrNull should return specified value for active and maxIssues
## Expected Behavior
`valueOrNull` delegates to originalConfig if value was defined

## Observed Behavior
`valueOrNull` does not delegate to originalConfig for `active` and for `maxIssues`

## Steps to Reproduce
ConfigurationsSpec -> should override active when specified -> switch to valueOrNull

## Context
I had a look at the assertions used in detekt to see if [Atrium](https://github.com/robstoll/atrium) would be a good fit. I'll create a WIP PR later on so that you can decide on your own.

## Your Environment
* Version of detekt used: master

### PR DESCRIPTION
**Pull request by robstoll**
- implemented valueOrDefault based on valueOrNull per default
- kept a different implementation for Test/YamlConfig

-fixes #2316

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt [hunk 1]
```diff
@@ -86,9 +86,10 @@ internal object EmptyConfig : HierarchicalConfig {
     override fun subConfig(key: String): EmptyConfig = this
 
     @Suppress("UNCHECKED_CAST")
-    override fun <T : Any> valueOrDefault(key: String, default: T): T = if (key == "active") true as T else default
-
-    override fun <T : Any> valueOrNull(key: String): T? = null
+    override fun <T : Any> valueOrNull(key: String): T? = when(key){
+        "active" -> true as? T
```
Comment (robstoll) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt]: maybe it should also cover `maxIssue` as in FailFastConfig?

**Comment by schalkms** [comment]
Thanks for tackling that! It looks good to me.

**Comment by robstoll** [comment]
@schalkms I have updated the documentation. Moreover, I had to change the PR slightly: 
ConfigAware and CompositeConfig need to override valueOrDefault due to YamlConfig and TestConfig which implement an additional behaviour in valueOrDefault (LSP violation right in your face :laughing:).
I don't know the reasons behind the parsing in valueOrDefault and whether this could be moved into Config as well (so that all configs behave the same way) or not.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt (hunk 1 above)
Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt]: Nah, maxIssues is now at 0 by default. The empty config just enables the rule under test.
I will move the BaseConfig and EmptyConfig to the internal package.

**Pull request merged**